Support CUDA tensors in ProcessGroupGloo

pietern opened the pull request:
Depends on #7628 being merged.

This also includes everything in #7628 because of this dependency. Once that is merged this will be rebased.

pietern committed:
Support CUDA tensors in ProcessGroupGloo

This adds an unconditional dependency on CUDA, which is not desirable
for the long term. Ideally we have split like ATen where we have
different artifacts for different backends so you can decide at runtime
what to use.

pietern commented:
I copied the AutoGPU struct from torch/csrcs/utils and ideally source it from some place instead of keeping a copy in this tree. It can also be AutoDevice instead of AutoGPU, depending on tensor type.

apaszke commented on torch/lib/c10d/CMakeLists.txt:
```diff
@@ -1,8 +1,52 @@
+  message(FATAL_ERROR "Gloo not found")
+endif()
+
+find_package(CUDA REQUIRED)
```
CUDA and gloo are required only for now, and we'll make them optional in the future, right?

pietern replied:
Yep, that's the plan. Don't yet know how, but I want to have backend specific libraries like ATen.

apaszke commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -297,6 +335,10 @@ EntryType ProcessGroupGloo::construct(const AlgorithmKey& key) {
   auto& srcSizes = key.srcSizes;
   entry->src.resize(srcSizes.size());
   for (int i = 0; i < srcSizes.size(); i++) {
+    std::unique_ptr<AutoGPU> gpu;
```
Why a `unique_ptr` and an extra allocation? It has a default constructor that doesn't do anything, better to use that.

pietern replied:
This is now fixed.

pietern committed:
Make ProcessGroupGloo::Work::finish take reference

pietern committed:
Remove unused work finish events

These were moved to the work class so that they have a lifecycle that is
independent from the algorithm entry instance.

pietern committed:
Fix cpu/cuda backend distinction in ProcessGroupGloo::Work

apaszke commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -43,8 +43,21 @@ class ProcessGroup {
+    // Ensures that operations on the output tensors that are invoked
+    // after this function returns are correctly sequenced after the
+    // asynchronous completion of this work.
+    virtual void synchronize() = 0;
```
Can you add some more docs for this function? Does it synchronize CUDA/CPU/both? Is there a description for the concurrency model of ProcessGroups somewhere? Can you refer to it from here? Would I ever want to call this manually instead of using `wait()`?

teng-li replied:
Yeah, i was gonna ask the same

pietern replied:
Added:

```
+    // For CUDA tensors, it inserts stream synchronization such that
+    // the streams of the caller wait for completion of the
+    // asynchronous operations on the destination tensors.
+    //
+    // For CPU tensors, it is currently a nop.
+    //
+    // This function should only be used if the caller polls for
+    // completion through the `isCompleted` function, it has returned
+    // true, and the `isSuccess` function also has returned true.
+    //
```

apaszke commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -84,9 +90,40 @@ const ::gloo::ReductionFunction<T>* reductionFunction(const ReduceOp& r) {
+
+// synchronizeStreams ensures that the private streams associated with
+// an algorithm entry wait for the public streams to complete, and
+// that the public streams wait for the private streams to complete.
```
This function only inserts public -> private synchronization, so the comment is misleading.

pietern replied:
Thanks for catching this. This was my original intent, but proved to be impossible, so changed it later.

apaszke commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2,9 +2,15 @@
 
+#include <THC.h>
+
+#include "private/CUDAUtils.hpp"
```
This means gloo will only be possible to compile when someone has CUDA. Do we really want to enforce that?

pietern replied:
For now, I'm fine with this.

I want to get a first rev out and making it depend on CUDA is reasonable IMO.

apaszke commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -339,13 +458,28 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::broadcast(
+  // In case of CUDA, ensure that operations that are queued after
+  // this collective wait for the collective to complete.
+  if (key.type->is_cuda()) {
+    synchronizeStreams(thcState_, entry);
```
Shouldn't this be part of the `run` body? Why do we only wait for public streams only once?

pietern replied:
We need to capture the state of the caller's streams at the call site. Everything in the run body is run by a different thread, where we have a different thread local streams.

apaszke commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -91,6 +95,27 @@ struct AlgorithmEntry {
+  // used to record completion of the memory copy of the input tensor.
+  //
+  std::vector<CUDAStream> streams;
+  std::vector<CUDAEvent> startEvents;
```
Are those events/streams E1/S1 or E2/S2 from the description above?

pietern replied:
Yes, updating the comment to be clearer.

apaszke commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -167,6 +184,13 @@ class ProcessGroupGloo : public ProcessGroup {
+    // asynchronously to finish this operation.
+    bool cuda_;
+    std::vector<int> devices_;
+    std::vector<CUDAEvent> events_;
```
Similar question like above. Would be good to have a reference to the note at the top.

pietern replied:
Added comment clarifying the use of local events here.

apaszke reviewed (commented):
@pietern @teng-li you might want to add yourself as CODEOWNERS of the c10d dir, because it will make your own approves unblock the merge.

teng-li commented on torch/lib/c10d/CMakeLists.txt:
```diff
@@ -23,18 +34,36 @@ else()
 
+find_package(CUDA REQUIRED)
+if(NOT CUDA_FOUND)
+  message(FATAL_ERROR "CUDA not found")
```
Do we want to fail the build if CUDA cannot be found? I am thinking of making it optional

pietern replied:
Right now I think it is acceptable, so we can get a first working version out the door.

pietern commented on torch/lib/c10d/CUDAUtils.hpp:
```diff
@@ -0,0 +1,103 @@
+    return *this;
+  }
+
+  cudaEvent_t getEvent() const {
```
It is const w.r.t. the CUDAEvent instance, but returns a non-const pointer.

pietern commented on torch/lib/c10d/CUDAUtils.hpp:
```diff
@@ -0,0 +1,103 @@
+  ~CUDADevice();
+
+ protected:
+  void setDevice(int device);
```
Good point, changing this.

pietern committed:
More comments

pietern committed:
Don't reset CUDA device between loop iterations

pietern commented:
@apaszke @teng-li The latest 2 commits should address your comments, can you check?

Looking into the CODEOWNERS thing now.

pietern committed:
Consolidate test helpers

teng-li commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -235,30 +305,68 @@ void ProcessGroupGloo::createAlgorithm(AlgorithmEntry& entry) {
+  }
+
+  if (backend == at::kCUDA) {
+    entry.algorithm = std::unique_ptr<::gloo::Algorithm>(
```
It would be nice to support GPU direct when creating the algorithm and when IB is available.  Can you create a TODO to automatically enable GPU direct when the device supports it?

pietern replied:
Added #8032

teng-li reviewed (approved):
Looks great!

teng-li commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -84,9 +90,40 @@ const ::gloo::ReductionFunction<T>* reductionFunction(const ReduceOp& r) {
+  CUDADevice deviceGuard;
+  const auto& key = entry->key;
+  for (auto i = 0; i < key.devices.size(); i++) {
+    const auto& device = key.devices[i];
```
maybe you can also do
```
deviceGuard.setDevice(device);
auto publicStream = THCState_getCurrentStream(thcState):
```

pietern replied:
I prefer the current version since it does 1 less cudaGetDevice call (even though it is cheap)

pietern commented:
@pytorchbot retest this please
